micrometer-core/src/main/java/io/micrometer/core/instrument/HighCardinalityTagsDetector.java

```diff
@@ -0,0 +1,156 @@
+        this(registry, threshold, delay, DEFAULT_CONSUMER);
+    }
+
+    /**
+     * @param registry The registry to use to check the Meters in it
+     * @param threshold The threshold to use to detect high cardinality tags (if the
+     * number of Meters with the same name are higher than this value, that's a high
+     * cardinality tag)
+     * @param delay The delay between the termination of one check and the commencement of
+     * the next
+     * @param meterNameConsumer The action to execute if the first high cardinality tag is
+     * found
+     */
+    public HighCardinalityTagsDetector(MeterRegistry registry, int threshold, Duration delay,
+            Consumer<String> meterNameConsumer) {
+        this.registry = registry;
+        this.threshold = threshold;
+        this.delay = delay;
+        this.meterNameConsumer = meterNameConsumer;
+        this.scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
+    }
+
+    /**
+     * Starts a scheduled job that checks if you have high cardinality tags.
+     */
+    public void start() {
+        this.scheduledExecutorService.scheduleWithFixedDelay(this::detectHighCardinalityTags, 0, this.delay.toMillis(),
+                TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * Shuts down the scheduled job that checks if you have high cardinality tags.
+     */
+    public void shutdown() {
+        this.scheduledExecutorService.shutdown();
+    }
+
+    @Override
+    public void close() {
+        this.shutdown();
+    }
+
+    private void detectHighCardinalityTags() {
+        try {
+            findFirst().ifPresent(this.meterNameConsumer);
+        }
+        catch (Exception exception) {
+            LOGGER.warn("Something went wrong during high cardinality tag detection", exception);
+        }
+    }
+
+    /**
+     * Finds the name of the first Meter that potentially has high cardinality tags.
+     * @return the name of the first Meter that potentially has high cardinality tags, an
+     * empty Optional if none found.
+     */
+    public Optional<String> findFirst() {
+        Map<String, Integer> meterNameFrequencies = new HashMap<>();
+        for (Meter meter : this.registry.getMeters()) {
+            String name = meter.getId().getName();
+            if (!meterNameFrequencies.containsKey(name)) {
+                meterNameFrequencies.put(name, 1);
+            }
+            else {
+                Integer frequency = meterNameFrequencies.get(name);
+                if (frequency < this.threshold) {
+                    meterNameFrequencies.put(name, frequency + 1);
+                }
+                else {
+                    return Optional.of(name);
+                }
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    private static void logWarning(String name) {
+        LOGGER.warn(String.format("It seems %s has high cardinality tags.", name));
```

Comment:
Do we want to continually log this at warn or would the warn-then-debug logger be better? I'm leaning toward voting the latter, but I could be convinced otherwise.